set new window parent

Edru2 opened the pull request:
<!-- To keep things simple, focus on just one topic in your PR and make changes just for that topic.
     This'll make it a lot easier to review, and thus your PR will get in faster.
     Remember can open multiple PRs at a time! -->
#### Brief overview of PR changes/additions

New lua ui function.
**setWindow**(windowname, name, [x, y, show])
Changes the window parent of the element (label, miniconsole, mapper) to the parent (userwindow) called windowname.
For mapper use the name "mapper".
To go back to the main window use the windowname "main".
the parameters x,y and show are optional.
x and y are set to 0 and show is true if not set.

new Geyser function :**changeContainer**(container)
Use it like:
testlabel:changeContainer(testcontainer2)
or
testemco:changeContainer(testuserwindow)
to change back to the main window (if your element wasn't in a container) use:
teslabel:changeContainer(Geyser)
...

Add the ability to Geyser to changes an elements container.
This works also if the container is in another parent window (userwindow)

to change the container and/or even the parent window.
#### Motivation for adding to Mudlet
Discussion on Discord and it was pretty simple.

#### Other info (issues closed, discussion etc)
All Geyser Elements should work. (plz test)
(now even flyout labels)

Edru2 committed:
set new window parent

New lua ui function.
setWindow(windowname, name, x, y)

new Geyser function :changeContainer
to change the container and/or even the parent window.

vadi2 commented on src/mudlet-lua/lua/geyser/GeyserGeyser.lua:
```diff
@@ -85,3 +85,32 @@ function Geyser:remove (window)
   table.remove(self.windows, index)
 end
+
+--- Removes a window from the parent it is in and put's it in a new one
```
```suggestion
--- Removes a window from the parent it is in and puts it in a new one
```

vadi2 commented on src/mudlet-lua/lua/geyser/GeyserGeyser.lua:
```diff
@@ -85,3 +85,32 @@ function Geyser:remove (window)
 end
+
+--- Removes a window from the parent it is in and put's it in a new one
+-- This is used internally, don't use it. Use changeContainer instead
```
They can't really, this function won't be available in `_G`

Edru2 committed:
added guiutils gauges, flyoutlabels, fix for showing hidden children

Added GUIutils gauges
Flyout labels
and before hidden children were shown if added to a container/userwindow
had to slightly change setWindow

vadi2 commented on src/TLuaInterpreter.cpp:
```diff
@@ -4366,9 +4367,21 @@ int TLuaInterpreter::setWindow(lua_State* L)
+    bool show = true;
+    if (n > 4) {
+        if ((!lua_isnumber(L, 5)) && (!lua_isboolean(L, 5))) {
+            lua_pushfstring(L, "setWindow: bad argument #5 type (show element as boolean/number (0/1) expected, got %s!)", luaL_typename(L, 5));
```
Let's not go with 0/1 for new functions - it was a mistake from the past

Edru2 committed:
Update GeyserGeyser.lua

Edru2 commented:
ready for testing :smiley:

vadi2 commented:
Awesome! Could you also get the people who wanted this to test?

vadi2 commented on src/TLuaInterpreter.cpp:
```diff
@@ -4332,6 +4332,64 @@ int TLuaInterpreter::moveWindow(lua_State* L)
+    }
+
+    if (!lua_isnumber(L, 3)) {
+        lua_pushfstring(L, "setWindow: bad argument #3 type (x-coordinate as number expected, got %s!)", luaL_typename(L, 3));
```
I think we can default to 0,0 for coordinates if not provided - a bit easier to use the api.

vadi2 commented on src/mudlet-lua/lua/GUIUtils.lua:
```diff
@@ -327,6 +327,19 @@ function showGauge(gaugeName)
+  show = show or true
+  assert(gaugesTable[gaugeName], "setGaugeWindow: no such gauge exists.")
+  assert(x and y, "setGaugeWindow: need to have both X and Y dimensions.")
+  setWindow(windowName, gaugeName .. "_back", x, y, show)
```
Same for defaults here.

vadi2 commented on src/mudlet-lua/lua/geyser/GeyserGeyser.lua:
```diff
@@ -85,3 +85,66 @@ function Geyser:remove (window)
+
+--- Removes a window from the container that it manages
+-- @param container The new container the window will be set in
+function Geyser:changeContainer (container)
```
Running this test throws up an error:

```lua
testuserwindow = Geyser.UserWindow:new({
  name = "TestUserWindow",
  x = "20%", y="20%",
  width="30%", height ="40%"
})

testlabel = Geyser.Label:new({
  name = "testlabel",
  x = "25%", y = "25%",
  width = "50%", height = "50%",
  fgColor = "black",
  message = [[<center>heey</center>]]
})
testlabel:setColor(0,255,0,150)

testlabel:changeContainer(testuserwindow)
testlabel:changeContainer("main")
```

![image](https://user-images.githubusercontent.com/110988/77139694-446f0380-6a77-11ea-8cb7-b90fc4d3afa2.png)

Edru2 replied:
This is not an error as changeContainer is not just a wrapper for setWindow but it is really changing the Container. You should put in a container which is on the main window or the Geyser Root Container Geyser. In this case
```lua
testlabel:changeContainer(Geyser)
```
But I will add an execption. If the string "main" is given container will just be Geyser.

And this error let me realize that I forgot to check if the container to change even exists (will add that).

Edru2 committed:
Added suggestions and proably fixed problem with label z axis

vadi2 commented:
It's without error now, but the `testlabel:setColor` has no effect from the code sample above now. Do you get the same?

Edru2 commented:
Copy pasted your example in Mudlet 4.6.2 
and in this PR and cannot see any difference in the label color.
So testlabel:setColor seems to work as expected for me.

Edru2 commented:
I have another strange issue.
The version for linux downloadable here doesn't seem to contain the last changes I made.
It still errors on your example.
This is the link I see:
https://make.mudlet.org/snapshots/d04567/Mudlet-4.6.1-testing-PR3480-3aa12c2-linux-x64.AppImage.tar

I tried to force reload the page and even to use another browser to download.

vadi2 commented:
Yes - see the `3aa12c2` in the link - it's for commit https://github.com/Mudlet/Mudlet/pull/3480/commits/3aa12c265a19b993e8ad75d4220e5ac86f794fc6. Somehow the bot didn't update the links, not sure why it doesn't happen often - but you can go to https://make.mudlet.org/snapshots and find it there anyway.

vadi2 commented:
Looks good to me, will let others review it too if they'd like!

Edru2 committed:
Prevent hidden container to get shown

vadi2 commented on src/mudlet-lua/lua/geyser/GeyserGeyser.lua:
```diff
@@ -74,7 +74,11 @@ function Geyser:add (window, cons)
     window:reposition()
   end
-  window:show()
+  if window.hidden or window.auto_hidden then
```
```suggestion
  if not (window.hidden or window.auto_hidden) then
```

?

Edru2 committed:
Update GeyserGeyser.lua

demonnic reviewed (approved):
functional testing passes, unit test output below

![image](https://user-images.githubusercontent.com/3660/77345460-05f17700-6d0b-11ea-8fc2-e6c9601c5d0c.png)
